src/complete.c

```diff
@@ -0,0 +1,258 @@
+  case vctrs_type_character: chr_detect_complete(proxy, size, p_out); break;
+  case vctrs_type_raw: raw_detect_complete(proxy, size, p_out); break;
+  case vctrs_type_list: list_detect_complete(proxy, size, p_out); break;
+  case vctrs_type_dataframe: df_detect_complete(proxy, size, p_out); break;
+  case vctrs_type_scalar: Rf_errorcall(R_NilValue, "Can't detect missing values in scalars with `vec_detect_complete()`.");
+  default: Rf_errorcall(R_NilValue, "Unimplemented type in `vec_detect_complete()`.");
+  }
+
+  UNPROTECT(2);
+  return out;
+}
+
+#define VEC_DETECT_COMPLETE(CTYPE, CONST_DEREF, SCALAR_EQUAL_MISSING) { \
+  const CTYPE* p_x = CONST_DEREF(x);                                    \
+                                                                        \
+  for (R_len_t i = 0; i < size; ++i) {                                  \
+    const CTYPE elt = p_x[i];                                           \
+    p_out[i] = !SCALAR_EQUAL_MISSING(elt);                              \
+  }                                                                     \
+}
+
+static inline
+void lgl_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(int, LOGICAL_RO, lgl_scalar_equal_missing);
+}
+static inline
+void int_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(int, INTEGER_RO, int_scalar_equal_missing);
+}
+static inline
+void dbl_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(double, REAL_RO, dbl_scalar_equal_missing);
+}
+static inline
+void cpl_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(Rcomplex, COMPLEX_RO, cpl_scalar_equal_missing);
+}
+static inline
+void chr_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(SEXP, STRING_PTR_RO, chr_scalar_equal_missing);
+}
+static inline
+void raw_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE(Rbyte, RAW_RO, raw_scalar_equal_missing);
+}
+
+#undef VEC_DETECT_COMPLETE
+
+#define VEC_DETECT_COMPLETE_BARRIER(SCALAR_EQUAL_MISSING) { \
+  for (R_len_t i = 0; i < size; ++i) {                      \
+    p_out[i] = !SCALAR_EQUAL_MISSING(x, i);                 \
+  }                                                         \
+}
+
+static inline
+void list_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  VEC_DETECT_COMPLETE_BARRIER(list_scalar_equal_missing);
+}
+
+#undef VEC_DETECT_COMPLETE_BARRIER
+
+// -----------------------------------------------------------------------------
+
+static inline void vec_detect_complete_col(SEXP x, R_len_t size, int* p_out);
+
+static
+void df_detect_complete(SEXP x, R_len_t size, int* p_out) {
+  // Initialize assuming fully complete
+  for (R_len_t i = 0; i < size; ++i) {
+    p_out[i] = 1;
+  }
+
+  int n_col = Rf_length(x);
+  const SEXP* p_x = VECTOR_PTR_RO(x);
+
+  for (R_len_t i = 0; i < n_col; ++i) {
+    const SEXP col = p_x[i];
+    vec_detect_complete_col(col, size, p_out);
```

Comment:
A bit strange to declare a const SEXP here.

```suggestion
    vec_detect_complete_col(p_x[i], size, p_out);
```